Etapa 4

munirwanis opened the pull request:
## Etapa 04

Modele a camada de **ViewModel** da aplicação. Nesse etapa, você faz com que o ViewModel consuma o fluxo de execução assíncrona, recebendo para isso o comportamento de domínio como um colaborador e usando as transformações adequadas no pipeline reativo para que a camada de visão no framework (UIViewController / Activity) renderize o resultado da forma adequada. 

Por exemplo, a interface pode consumir  alguns estados bem definidos de acordo com as emissões, a saber :

- Comece a tela em `WAITING_FOR_INPUT`
- Passe para o estado `LOADING`
- Receba os itens processados para renderizar na listagem (`SUCCESS`)
- Ou, alternativamente, recebe um indicador para o estado `ERROR` (erro esse previamente modelado pelo que você já tratou na infra-estrutura de dados)

Além disso, considere também nessa etapa que na UI (etapa a seguir) você precisará que os itens da listagem sigam os seguintes critérios visuais

- Fatos com mais de 50 caracteres deveram ser mostrados com fonte 14; senão fonte 16
- Quando um fato tem uma categoria associada, seu label é AZUL, claro contrário, é CINZA


Você deve cobrir esse ViewModel (e eventuais colaboradores) com testes que garantam que

- A interação com a fonte de dados ocorre como se espera
- As transformações de containers para dados (domínio -> visão) ocorre como se espera
- A sequência de emissões para a interface ocorre como se espera através do pipeline reativo

munirwanis commented:
Ignorem alterações na ViewController, não faz parte da etapa.

mariimeireles commented on ChuckNorrisFacts/Presentation/List Facts/ListFactsState.swift:
```diff
@@ -0,0 +1,27 @@
+}
+
+
+extension ListFactsState: Equatable {
```
a extension para Equatable é só para testes?

munirwanis replied:
Sim, Tati também comentou, vou mudar isso e fazer um teste melhor

mariimeireles commented on ChuckNorrisFacts/Presentation/List Facts/ListFactsViewModel.swift:
```diff
@@ -9,18 +9,38 @@
+// MARK: - Private methods
+
+private extension ListFactsViewModel {
+    private func convert(_ facts: Facts) -> FactsPresentation {
```
e se a busca não retornar nenhum resultado? não precisa checar se está vazio?

tatimagdalena replied:
Esse método pode ficar num mapper separado também!

tatimagdalena commented on ChuckNorrisFactsTests/View Model Tests/ListFactsViewModelTests.swift:
```diff
@@ -0,0 +1,182 @@
+import Nimble
+import Quick
+import RxSwift
+import RxTest
```
Seria interessante testar nesse PR o RxBlocking, já que vai ser o último com testes unitários.
Só pra você conseguir sentir a diferença de abordagens :)

tatimagdalena commented on ChuckNorrisFacts/Presentation/List Facts/ListFactsState.swift:
```diff
@@ -0,0 +1,27 @@
+}
+
+
+extension ListFactsState: Equatable {
```
Colocar o equatable em enum no código de produção só pra ser usado em assert de teste pode ser uma dor de cabeça, porque quando o enum tem tipos associados você tem que ir aplicando o Equatable em toda a árvore de estruturas...
Uma abordagem que ajuda nisso pode ser vista [nesse artigo](https://qualitycoding.org/swift-equatable-testing).
Ele usa o switch/case do enum, e caso não caia no que você quer, dá o fail. Fica bem mais simples :)

CalebeEmerick commented on ChuckNorrisFacts/Presentation/List Facts/ListFactsViewModel.swift:
```diff
@@ -9,18 +9,38 @@
+final class ListFactsViewModel {
+    private let service: FactServiceProtocol
+    private let bag = DisposeBag()
+    var currentState = Variable<ListFactsState>(.waitingForInput)
```
Variable tá `deprecated`, será oficialmente depreciado no próximo grande release. Depois dá uma olhada nas alternativas.

munirwanis replied:
![Hello Darkness](https://media3.giphy.com/media/3o7WTqo27pLRYxRtg4/giphy.gif)

CalebeEmerick commented on ChuckNorrisFactsTests/View Model Tests/ListFactsViewModelTests.swift:
```diff
@@ -0,0 +1,182 @@
+import RxSwift
+import RxTest
+
+class ListFactsViewModelTests: QuickSpec {
```
Não acha que dessa forma dificulta bastante a legibilidade?

munirwanis replied:
Acho muito pelo contrário!

tatimagdalena replied:
Um problema que eu vejo de fazer assim é a impossibilidade de rodar um teste por vez...

munirwanis replied:
Como a gente estava discutindo aqui pessoalmente outro dia (vou colocar aqui só pra ficar registrado), eu resolvi usar ao máximo o framework de `Spec` mais pra conhecer e ver o que da pra fazer. 

Não pretendo usar no dia-a-dia porque sabemos o inferno que é (enquanto o Swift não for API-stable) quando rolam updates da linguagem e as coisas começam a quebrar. Eu ainda prefiro usar o `XCTest` porque é o default da Apple e funciona.

Como eu mexi com Ruby uma época pra mim é muito clara a escrita dos testes com o `Spec`, mas entendo que seja meio ruim pra galera que já tá acostumada com o `XCTest`.

munirwanis committed:
🔥  Removes carthage files

munirwanis committed:
# This is a combination of 3 commits.
# This is the 1st commit message:

👊  Resolves conflicts

# This is the commit message #2:

🔥 Removes timeout from image downloader observer

# This is the commit message #3:

✅ Adds more tests to ViewModel

munirwanis committed:
# This is a combination of 4 commits.
# This is the 1st commit message:

👊 Resolves README conflicts

# This is the commit message #2:

🎨 Changes Variable BehaviorRelay since Variable is deprecated

# This is the commit message #3:

✅  Refactor more tests to use RxBlocking

# This is the commit message #4:

✅ Changes tests to use RxBlocking

munirwanis committed:
👊  Resolves conflicts with tests

munirwanis committed:
Etapa 4

Modele a camada de **ViewModel** da aplicação. Nesse etapa, você faz com que o ViewModel consuma o fluxo de execução assíncrona, recebendo para isso o comportamento de domínio como um colaborador e usando as transformações adequadas no pipeline reativo para que a camada de visão no framework (UIViewController / Activity) renderize o resultado da forma adequada.

Por exemplo, a interface pode consumir  alguns estados bem definidos de acordo com as emissões, a saber :

- Comece a tela em `WAITING_FOR_INPUT`
- Passe para o estado `LOADING`
- Receba os itens processados para renderizar na listagem (`SUCCESS`)
- Ou, alternativamente, recebe um indicador para o estado `ERROR` (erro esse previamente modelado pelo que você já tratou na infra-estrutura de dados)

Além disso, considere também nessa etapa que na UI (etapa a seguir) você precisará que os itens da listagem sigam os seguintes critérios visuais

- Fatos com mais de 50 caracteres deveram ser mostrados com fonte 14; senão fonte 16
- Quando um fato tem uma categoria associada, seu label é AZUL, claro contrário, é CINZA

Você deve cobrir esse ViewModel (e eventuais colaboradores) com testes que garantam que

- A interação com a fonte de dados ocorre como se espera
- As transformações de containers para dados (domínio -> visão) ocorre como se espera
- A sequência de emissões para a interface ocorre como se espera através do pipeline reativo

👊 Resolves conflicts with pbxproj

✅  Adds test to cover empty state

📝  Adds RxBlocking articles

munirwanis committed:
🔥  Removes Log call
